### PR TITLE
8318678: Vector access on heap MemorySegments only works for byte[]

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes removing the restriction that only heap `MemorySegment` wrapping a `byte` array can be accessed by Vectors. Now any array type can be used provided the element alignment constraints are respected.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8318956](https://bugs.openjdk.org/browse/JDK-8318956) to be approved

### Issues
 * [JDK-8318678](https://bugs.openjdk.org/browse/JDK-8318678): Vector access on heap MemorySegments only works for byte[] (**Bug** - P3)
 * [JDK-8318956](https://bugs.openjdk.org/browse/JDK-8318956): Vector access on heap MemorySegments only works for byte[] (**CSR**)

### Reviewers
 * [Chris Hegarty](https://openjdk.org/census#chegar) (@ChrisHegarty - **Reviewer**) ⚠️ Review applies to [4828a98b](https://git.openjdk.org/jdk/pull/16360/files/4828a98b829f98334e6871538a3bbd02d703780b)
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [4828a98b](https://git.openjdk.org/jdk/pull/16360/files/4828a98b829f98334e6871538a3bbd02d703780b)
 * [Paul Sandoz](https://openjdk.org/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to [c104051c](https://git.openjdk.org/jdk/pull/16360/files/c104051c3b9c33cd9165c0473c7449dc8158991e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16360/head:pull/16360` \
`$ git checkout pull/16360`

Update a local copy of the PR: \
`$ git checkout pull/16360` \
`$ git pull https://git.openjdk.org/jdk.git pull/16360/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16360`

View PR using the GUI difftool: \
`$ git pr show -t 16360`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16360.diff">https://git.openjdk.org/jdk/pull/16360.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16360#issuecomment-1779253424)

### CODE REVIEW

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/AbstractVector.java [hunk 1]
```diff
@@ -736,6 +742,25 @@ AbstractVector<F> convert0(char kind, AbstractSpecies<F> rsp) {
         throw new AssertionError();
     }
 
+    /*package-private*/
+    @ForceInline
+    final
+    AbstractMemorySegmentImpl requireSegmentConvertibleFor(MemorySegment segment, long offset, int elementByteSize)  {
+        AbstractMemorySegmentImpl ams = (AbstractMemorySegmentImpl) segment;
+        if (ams.maxAlignMask() > 1 && !ams.isAlignedForElement(offset, elementByteSize)) {
```
Comment (mcimadamore) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/AbstractVector.java:750]: I don't think we should only check for maxAlignMask > 1 - you also want to check native segments, right?

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java [hunk 2]
```diff
@@ -3282,7 +3281,8 @@ ByteVector fromBooleanArray(VectorSpecies<Byte> species,
      *         or {@code offset+N*1 >= ms.byteSize()}
      *         for any lane {@code N} in the vector
      * @throws IllegalArgumentException if the memory segment is a heap segment that is
-     *         not backed by a {@code byte[]} array.
+     *         not backed by a {@code byte[]} array and if access to the backing array
```
Comment (mcimadamore) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java:3284]: This seems backwards - e.g. a ByteVector requires an alignment of 1. So I would expect to be able to pass _any_ segment?

File: test/jdk/jdk/incubator/vector/AbstractVectorLoadStoreTest.java [hunk 3]
```diff
@@ -50,13 +54,96 @@ public class AbstractVectorLoadStoreTest extends AbstractVectorTest {
...
     );
 
+    static Stream<IntFunction<MemorySegment>> memorySegmentGenerators(ValueLayout elementLayout) {
+        return MEMORY_SEGMENT_GENERATORS.stream()
+                .filter(f -> canBeConverted(f, elementLayout));
+    }
+
+    private static boolean canBeConverted(IntFunction<MemorySegment> function, ValueLayout elementLayout) {
+        // Create a sample to analyze
+        MemorySegment s = function.apply(Long.BYTES);
```
Comment (mcimadamore) [inline test/jdk/jdk/incubator/vector/AbstractVectorLoadStoreTest.java:118]: I believe that a good way to test this is the following:
* each vector type operates as having a given element layout - for instance, you can imagine the layout for IntVector to be `JAVA_INT` and so forth
* asking whether you load a segment into a vector is the same as asking whether you can access the segment, at offset 0L, with the layout associated with the vector (see above) - that is if MS::get throws, then vector load should also throw (and viceversa)

Comment (ChrisHegarty) [reply test/jdk/jdk/incubator/vector/AbstractVectorLoadStoreTest.java:118]: Vectors use and have in their docs, a layout with a byte alignment of 1. e.g. from `IntVector`

```
 ValueLayout.OfInt ELEMENT_LAYOUT = ValueLayout.JAVA_INT.withByteAlignment(1)
```

Can the `fromMemorySegment` not just behave similarly? I get that alignment is preferable, but does it need to be enforced?  If so, then maybe the `ELEMENT_LAYOUT` and example in the javadoc needs to be updated.

**Comment by merykitty** [comment]
Vector loads and stores are always unaligned, so I don't see the value in forcing the alignment of accesses.

File: test/jdk/jdk/incubator/vector/AbstractVectorLoadStoreTest.java (hunk 3 above)
Comment (minborg) [reply test/jdk/jdk/incubator/vector/AbstractVectorLoadStoreTest.java]: I've removed the alignment checks.

**Review by ChrisHegarty — APPROVED** [review]
LGTM

**Comment by mcimadamore** [comment]
On a second look, I agree with the comments. Logically, the vector API is accessing elements using `JAVA_XYZ_UNALIGNED` layouts, which should _always_ succeed. E.g. no alignment exception should ever be possible when doing unaligned access, regardless of whether the segment is on/off heap.

**Review by mcimadamore — APPROVED** [review]

**Comment by PaulSandoz** [comment]
We also need to review the intrinsic for load/store found here to ensure, when the hardware supports it, that mixed access is optimized:

https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vectorIntrinsics.cpp#L967

I recommend writing a simple benchmark with various forms of mixed access and verify the intrinsics kick (longer term IR tests would be preferable).

**Comment by minborg** [comment]
Here is a test showing there is a need to improve performance for certain combinations of load operations. It is likely, the same is true for store operations.

```
Benchmark                                                   (size)  Mode  Cnt     Score     Error  Units
TestLoadSegmentVarious.byteVectorFromByteBackedSegment        1024  avgt   10   280.008 ?   7.251  ns/op
TestLoadSegmentVarious.byteVectorFromDoubleBackedSegment      1024  avgt   10  1304.008 ?  98.901  ns/op
TestLoadSegmentVarious.byteVectorFromIntBackedSegment         1024  avgt   10  1279.621 ? 100.008  ns/op
TestLoadSegmentVarious.doubleVectorFromByteBackedSegment      1024  avgt   10    37.281 ?   1.360  ns/op
TestLoadSegmentVarious.doubleVectorFromDoubleBackedSegment    1024  avgt   10    36.847 ?   0.130  ns/op
TestLoadSegmentVarious.doubleVectorFromIntBackedSegment       1024  avgt   10   194.195 ?  31.096  ns/op
TestLoadSegmentVarious.intVectorFromByteBackedSegment         1024  avgt   10    72.602 ?   1.768  ns/op
TestLoadSegmentVarious.intVectorFromDoubleBackedSegment       1024  avgt   10   166.851 ?   9.528  ns/op
TestLoadSegmentVarious.intVectorFromIntBackedSegment          1024  avgt   10    71.283 ?   0.507  ns/op
TestLoadSegmentVarious.scalarByteVectorFromByteSegment        1024  avgt   10  4790.084 ?  45.882  ns/op
TestLoadSegmentVarious.scalarByteVectorFromDoubleSegment      1024  avgt   10  4841.273 ? 291.962  ns/op
TestLoadSegmentVarious.scalarByteVectorFromIntSegment         1024  avgt   10  4794.028 ? 101.282  ns/op
TestLoadSegmentVarious.scalarDoubleVectorFromByteSegment      1024  avgt   10  1241.117 ?  11.603  ns/op
TestLoadSegmentVarious.scalarDoubleVectorFromDoubleSegment    1024  avgt   10  1245.752 ?  15.516  ns/op
TestLoadSegmentVarious.scalarDoubleVectorFromIntSegment       1024  avgt   10  1232.216 ?   8.365  ns/op
TestLoadSegmentVarious.scalarIntVectorFromByteSegment         1024  avgt   10  1239.146 ?  14.582  ns/op
TestLoadSegmentVarious.scalarIntVectorFromDoubleSegment       1024  avgt   10  1236.712 ?   8.063  ns/op
TestLoadSegmentVarious.scalarIntVectorFromIntSegment          1024  avgt   10  1228.656 ?   3.329  ns/op
```

File: test/jdk/jdk/incubator/vector/IntrinsicHeapTest.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+import java.lang.foreign.ValueLayout;
+import java.nio.ByteOrder;
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.LockSupport;
+import java.util.stream.IntStream;
+
+import static org.testng.Assert.*;
+
+public class IntrinsicHeapTest {
```
Comment (PaulSandoz) [inline test/jdk/jdk/incubator/vector/IntrinsicHeapTest.java:48]: I don't think we need this test as part of this PR. It's useful ascertain what is intrinsic or not. We can rethink it as part of the following fix if needed to verify the intrinsics work across all the cross product of array types and vector types.

**Review by PaulSandoz — APPROVED** [review]
Looks good. Please remove `IntrinsicHeapTest` and let's follow up in another issue/PR to fix the mismatched heap access not being intrinsic.

**Pull request closed without merging**